### ISSUE TITLE
libzt incorrectly passes Winsock-style structures into lwIP directly

### ISSUE THREAD

**Issue by hach-que** [zerotier/libzt#26]
So one of the issues with libzt on Windows is that `libzt.h` includes the Winsock headers, but the Windows structures for types like `fd_set` and `sockaddr_in` are incompatible with the lwIP structures. However, libzt [just passes the pointer through to lwIP](https://github.com/zerotier/libzt/blob/c00f2a97e6f9ffa24ad42e694992edd26227304f/src/libzt.cpp#L387) [assuming the structure is the same](https://github.com/zerotier/libzt/blob/c00f2a97e6f9ffa24ad42e694992edd26227304f/src/libzt.cpp#L531), and this causes things to break in subtle ways.

It means we're currently having to do workarounds like this in order to actually use `zts_*` functions:

https://github.com/HiveMP/ztserver/blob/semaphore-broken/forwardzt.cpp#L11-L31
https://github.com/HiveMP/ztserver/blob/semaphore-broken/forwardzt.cpp#L181 (this is a forced cast because the Winsock sockaddr_in6 and the lwIP sockaddr_in6 are colliding)

This impacts functions like `zts_select` which use `fd_set` and `zts_recvfrom` which write the received address back into a `sockaddr` pointer.

For example, this is the Windows (Winsock) definition of fd_set:

```
typedef struct fd_set {
        u_int fd_count;               /* how many are SET? */
        SOCKET  fd_array[FD_SETSIZE];   /* an array of SOCKETs */
} fd_set;
```

And this is the lwIP definition of fd_set:

```
typedef struct fd_set
{
  unsigned char fd_bits [(FD_SETSIZE+7)/8];
} fd_set;
```

This also impacts macros like `FD_ZERO`, `FD_SET` and `FD_ISSET` which are defined differently as well.

This is the Windows (Winsock) definition of `struct sockaddr_in6` (used in `zts_recvfrom`):

```
typedef struct sockaddr_in6 {
    ADDRESS_FAMILY sin6_family; // AF_INET6.
    USHORT sin6_port;           // Transport level port number.
    ULONG  sin6_flowinfo;       // IPv6 flow information.
    IN6_ADDR sin6_addr;         // IPv6 address.
    union {
        ULONG sin6_scope_id;     // Set of interfaces for a scope.
        SCOPE_ID sin6_scope_struct;
    };
} SOCKADDR_IN6_LH, *PSOCKADDR_IN6_LH, FAR *LPSOCKADDR_IN6_LH;
```

Yet this is the lwIP definition of `struct sockaddr_in6`:

```
struct sockaddr_in6 {
  u8_t            sin6_len;      /* length of this structure    */
  sa_family_t     sin6_family;   /* AF_INET6                    */
  in_port_t       sin6_port;     /* Transport layer port #      */
  u32_t           sin6_flowinfo; /* IPv6 flow information       */
  struct in6_addr sin6_addr;     /* IPv6 address                */
  u32_t           sin6_scope_id; /* Set of interfaces for scope */
};
```

Even the definition of `struct in6_addr` varies in terms of accessing the data between Windows:

```
typedef struct in6_addr {
    union {
        UCHAR       Byte[16];
        USHORT      Word[8];
    } u;
} IN6_ADDR, *PIN6_ADDR, FAR *LPIN6_ADDR;
```

and lwIP:

```
struct ip6_addr {
  u32_t addr[4];
};
```

Anyway, you get the point. The difference in data structures causes everything to misbehave when data is being passed between the libzt APIs (using Winsock) and the underlying lwIP code.

I think the correct fix here is either:
- libzt needs to convert the data structures back and forth between formats in libzt.cpp when on Windows (hard because all the types conflict on names)
- libzt needs to use lwip/sockets.h instead of Winsock for the external facing API on Windows

This issue probably doesn't occur on Linux/Mac because the lwIP structures probably match what `sys/socket.h` defines.

**Comment by hach-que** [comment]
Also of note, this probably wasn't picked up in regular testing because the most common structures and methods (`struct sockaddr_in` and `zts_read`) don't have these kinds of incompatibilities.

`zts_read` doesn't return any sender information (it just returns the byte array as a pointer and doesn't try to fill in a `sockaddr` struct).

`struct sockaddr_in` (i.e. IPv4) is _mostly_ runtime compatible between Winsock:

```
typedef struct sockaddr_in {

#if(_WIN32_WINNT < 0x0600)
    short   sin_family;
#else //(_WIN32_WINNT < 0x0600)
    ADDRESS_FAMILY sin_family;
#endif //(_WIN32_WINNT < 0x0600)

    USHORT sin_port;
    IN_ADDR sin_addr;
    CHAR sin_zero[8];
} SOCKADDR_IN, *PSOCKADDR_IN;
```

and lwIP:

```
struct sockaddr_in {
  u8_t            sin_len;
  sa_family_t     sin_family;
  in_port_t       sin_port;
  struct in_addr  sin_addr;
#define SIN_ZERO_LEN 8
  char            sin_zero[SIN_ZERO_LEN];
};
```

The `sin_len` (8 bits) and `sin_family` (8 bits) fields in the lwIP structure map over the `sin_family` (16 bits) field in the Winsock structure, causing `sin_port` and `sin_addr` to line up correctly.

So you'd only notice the issue with IPv4 addresses if you tried to read `sin_family`; `sin_port` and `sin_addr` will be correct (and unlike most test code, in `ztserver` we actually do use the family field because we receive/send both IPv4 and IPv6 packets over ZeroTier).

**Comment by joseph-henry** [comment]
Thanks for putting this together. This is something I'd like libzt to handle so the user/app is isolated from this difference in implementation. I'll soon make some changes to the `edge` branch to address this.

**Comment by joseph-henry** [comment]
This is an old issue but one that I have deliberately not closed because I hadn't actually resolved it. However, there have been a few changes to the project since the opening of this ticket which might resolve this issue and I'm curious as to whether it still exists:

 - With the latest version I've provided better integration with the underlying network stack's functionality and exposed more of its native structures to the user via `ZeroTierSockets.h`. Previously I provided a sort of half-baked compatibility layer which wasn't very reliable. Now, if the user application sticks to using `zts_*` structures and `ZTS_*` constants defined in the C API header there's a lower chance of anything getting interpreted incorrectly.

- I've updated the lwIP submodule to 2.1.2 STABLE

If this remains a bug I am willing to re-approach it but it may be a bug/oversight in the lwIP code.

Let me know what help if any you'd like on this. And thanks again for your time and effort on this!
